### PR TITLE
Refactor options management - Part 3

### PR DESCRIPTION
**Pull request by iMoses**

### CODE REVIEW

File: packages/ag-charts-community/src/chart/__image_snapshots__/gallery-examples-test-ts-gallery-examples-ag-chart-v-2-create-for-bubble-graph-with-categories-example-it-should-render-to-canvas-as-expected-1-snap.png
Comment (alantreadway) [inline packages/ag-charts-community/src/chart/__image_snapshots__/gallery-examples-test-ts-gallery-examples-ag-chart-v-2-create-for-bubble-graph-with-categories-example-it-should-render-to-canvas-as-expected-1-snap.png:1]: Should axis-lines be visible for this example?

File: packages/ag-charts-community/src/chart/__image_snapshots__/gallery-examples-test-ts-gallery-examples-ag-chart-v-2-create-for-cross-lines-example-it-should-render-to-canvas-as-expected-1-snap.png
Comment (alantreadway) [inline packages/ag-charts-community/src/chart/__image_snapshots__/gallery-examples-test-ts-gallery-examples-ag-chart-v-2-create-for-cross-lines-example-it-should-render-to-canvas-as-expected-1-snap.png:1]: Crossline ranges have changed colour, is that expected?

File: packages/ag-charts-community/src/chart/agChartV2.ts [hunk 1]
```diff
@@ -390,7 +392,7 @@ class AgChartsInternal {
 }
 
 function applyChartOptions(chart: Chart, processedOptions: ProcessedOptions, userOptions: AgChartOptions): void {
-    const completeOptions = jsonMerge([chart.processedOptions ?? {}, processedOptions], noDataCloneMergeOptions);
+    const completeOptions = mergeDefaults(processedOptions, chart.processedOptions);
```
Comment (alantreadway) [inline packages/ag-charts-community/src/chart/agChartV2.ts:395]: Is the precedence here correct? The new options state (`processedOptions`) should take precedence over the previously determined options state (`chart.processedOptions`)?

File: packages/ag-charts-community/src/chart/agChartV2.ts [hunk 2]
```diff
@@ -443,7 +445,7 @@ function applyChartOptions(chart: Chart, processedOptions: ProcessedOptions, use
         chart.updateAllSeriesListeners();
     }
     chart.processedOptions = completeOptions;
-    chart.userOptions = jsonMerge([chart.userOptions ?? {}, userOptions], noDataCloneMergeOptions);
+    chart.userOptions = mergeDefaults(userOptions, chart.userOptions);
```
Comment (alantreadway) [inline packages/ag-charts-community/src/chart/agChartV2.ts:448]: Likewise precedence.

File: packages/ag-charts-community/src/chart/factory/chartTypes.ts [hunk 3]
```diff
@@ -40,7 +40,7 @@ export function registerChartSeriesType(seriesType: string, chartType: ChartType
 }
 
 export function registerChartDefaults(chartType: ChartType, defaults: {}) {
-    DEFAULTS[chartType] = jsonMerge([DEFAULTS[chartType] ?? {}, defaults]);
+    DEFAULTS[chartType] = mergeDefaults(defaults, DEFAULTS[chartType]);
```
Comment (alantreadway) [inline packages/ag-charts-community/src/chart/factory/chartTypes.ts:43]: Precedence?

File: packages/ag-charts-community/src/chart/mapping/defaults.ts [hunk 4]
```diff
@@ -32,26 +29,3 @@ export function swapAxes<T extends AgChartOptions>(opts: T): T {
         ],
     };
 }
-
-type PossibleObject = { enabled?: boolean } | undefined;
-
-export function resolveModuleConflicts<T extends AgChartOptions>(opts: T) {
```
Comment (alantreadway) [inline packages/ag-charts-community/src/chart/mapping/defaults.ts:38]: Where has this check moved to?

**Review by alantreadway — COMMENTED** [review]
<partial review>

File: packages/ag-charts-community/src/chart/navigator/__image_snapshots__/navigator-test-ts-navigator-create-for-nav-zoomed-edge-crosslines-1-it-should-render-to-canvas-as-expected-1-snap.png
Comment (alantreadway) [inline packages/ag-charts-community/src/chart/navigator/__image_snapshots__/navigator-test-ts-navigator-create-for-nav-zoomed-edge-crosslines-1-it-should-render-to-canvas-as-expected-1-snap.png:1]: Chart padding change expected?

File: packages/ag-charts-community/src/chart/series/__image_snapshots__/series-labels-test-ts-series-labels-create-for-bubble-series-labels-it-should-render-to-canvas-as-expected-1-snap.png
Comment (alantreadway) [inline packages/ag-charts-community/src/chart/series/__image_snapshots__/series-labels-test-ts-series-labels-create-for-bubble-series-labels-it-should-render-to-canvas-as-expected-1-snap.png:1]: Axis lines now visible?

File: packages/ag-charts-community/src/util/json.test.ts [hunk 5]
```diff
@@ -197,40 +198,29 @@ describe('json module', () => {
...
+                const mergee1: any = { a: [] };
+                const mergee2: any = { a: [[{ x2: 1 }, { y2: 1 }], [{ m2: 2, n2: 2 }]] };
 
-                const merge = jsonMerge([base, mergee1, mergee2]);
+                const merge = mergeDefaults(mergee2, mergee1, base);
                 expect(merge).toMatchSnapshot();
                 expect(merge).toEqual(mergee2);
             });
 
-            it('should honour DELETE value', () => {
```
Comment (alantreadway) [inline packages/ag-charts-community/src/util/json.test.ts:217]: This case was to support users unsetting options on update, are you happy that this scenario is catered for?

File: packages/ag-charts-enterprise/src/axes/__image_snapshots__/polar-axes-test-ts-polar-axes-should-render-radius-cross-line-range-as-expected-1-snap.png
Comment (alantreadway) [inline packages/ag-charts-enterprise/src/axes/__image_snapshots__/polar-axes-test-ts-polar-axes-should-render-radius-cross-line-range-as-expected-1-snap.png:1]: Crossline rendering change expected?

**Review by alantreadway — COMMENTED** [review]
All reviewed except for skipped/commented tests.

File: packages/ag-charts-community/src/chart/mapping/defaults.ts (hunk 4 above)
Comment (iMoses) [reply packages/ag-charts-community/src/chart/mapping/defaults.ts:38]: https://github.com/ag-grid/ag-charts/pull/842/files/c9663906e8344ec265cc2b1ec00cedf3b2998dba#diff-08278e1c316b3be34805c99a6ac7216e4dc5a2d551ec152a508da7a58346e2c7R494

File: packages/ag-charts-community/src/util/json.test.ts (hunk 5 above)
Comment (iMoses) [reply packages/ag-charts-community/src/util/json.test.ts:217]: It was used only to remove properties from results and wasn't exported to the users but only internally.
I personally think it was an additional mechanism that just added unnecessary complexity, instead I just "manually" ignore specific properties in the 2-4 places where it was required.

**Review by alantreadway — APPROVED** [review]
LGTM

**Pull request merged**